spinalcordtoolbox/scripts/sct_apply_transfo.py

```diff
@@ -109,6 +111,44 @@ def get_parser():
     return parser
 
 
+def isct_antsApplyTransforms(dimensionality,
+                             fname_input,
+                             fname_output,
+                             fname_reference,
+                             transforms,
+                             interpolation_args,
+                             verbose):
+    """
+    Wrapper for CLI binary that allows for pre/postprocessing steps.
+
+    Note: Most logic should still go in `sct_apply_transfo`, because our SCT
+    script is essentially one big wrapper script for the ANTs call. Only use
+    this wrapper to fix bugs in the binary itself, or to compensate for
+    missing behavior/options in the ANTs source code. This wrapper should more
+    or less function like the original binary (with only minor tweaks added).
+    """
+    dtype_in = Image(fname_input).data.dtype
+    run_proc(['isct_antsApplyTransforms',
+              '-d', dimensionality,
+              '-i', fname_input,
+              '-o', fname_output,
+              '-t'] + transforms +  # list of transforms
+             ['-r', fname_reference] +
+             interpolation_args,  # list of ['-n', interp_type]
+             verbose=verbose,
+             is_sct_binary=True)
+    # Preserve input datatype if output data can be safely cast to it
+    # Note: This is mostly relevant for integer input images + NearestNeighbour
+    #       interpolation, since ANTs will automatically cast to float, but
+    #       we usually want to preserve the integer datatype.
+    im_out = Image(fname_output)
+    if np.array_equal(im_out.data, im_out.data.astype(dtype_in)):
```

Comment:
I would have expected this `if` statement to do a static check for certain arguments in the call to `isct_antsAppleTransforms` (something like `'nn' in interpolation_args`), rather than a dynamic check of whether the actual data values are affected by the casting operation. Is there a case that's better handled this way?

I ask because my experience is that having the output _type_ change based on the output _values_ can lead to very hard to debug corner cases.

Reply:
Thank you for the feedback! It definitely makes sense to be explicit about the cases we wish to apply this logic too.

I think the reason I wrote it this way was because I figured it was a safe operation to apply blindly (since we still ensure array equality), and we're applying this operation to a file that we generate (meaning the file is kind of still within "our domain")? And I figured it might help us catch other instances of type changing that we can't think of on the spot (e.g. the user supplies a float32 file or something because they're keen on saving disk space, but ANTs outputs a float64 file by default, etc.)

But, I'm alright with reverting my logic and explicitly defining the cases we want to apply this logic to (as to match the scope of the original issue).